### ISSUE TITLE
Negative values net change

### ISSUE THREAD

**Issue by mhesselbarth** [renatoamorais/rfishprod#2]
Hey, 

First of all thanks for contributing this very cool and interesting package, well done! 

When going through the provided example, I noticed that the net change (i.e., somatic growth minus somatic loss) is negative for quite a few individuals (see example code). Is there something obvious I am missing or doing wrong? 

Thanks a lot. 

``` r
library(rfishprod)
#> Loading required package: xgboost
#> This is rfishprod v0.0.1. Please, note that this is still a beta release.

# get example data
input_data <- rfishprod:::repdata
  
# clean trait factors
input_data_cln <- rfishprod::tidytrait(input_data, rfishprod::db)

# formula from Morais and Bellwood (2018) 
fmod <- formula(~ sstmean + MaxSizeTL + Diet + Position + Method)

# predict Kmax and only save prediction
pred_data <- rfishprod::predKmax(traits = input_data_cln, dataset = db,
                            fmod = fmod, niter = 10)
#> Bootstrapping the model, round 1 
#> Bootstrapping the model, round 2 
#> Bootstrapping the model, round 3 
#> Bootstrapping the model, round 4 
#> Bootstrapping the model, round 5 
#> Bootstrapping the model, round 6 
#> Bootstrapping the model, round 7 
#> Bootstrapping the model, round 8 
#> Bootstrapping the model, round 9 
#> Bootstrapping the model, round 10

pred_data <- pred_data$pred

# predict mortality rates
pred_data$Mort <- rfishprod::predM(Lmeas = pred_data$Size, Lmax = pred_data$MaxSizeTL,
                                   Kmax = pred_data$Kmax,  method = "Gislason")

# calculate fish growth next timestep and net increase
pred_data$Size_t1 <- rfishprod::applyVBGF(Lmeas = pred_data$Size,
                                          Lmax = pred_data$MaxSizeTL,
                                          Kmax = pred_data$Kmax)

pred_data$Som_gain_a <- rfishprod::somaGain(a = pred_data$a, b = pred_data$b,
                                            Lmeas = pred_data$Size, Lmax = pred_data$MaxSizeTL,
                                            Kmax = pred_data$Kmax)

# calculate mass gain "by hand"
pred_data$Mass <- pred_data$a * (pred_data$Size ^ pred_data$b) #mass before growth (t0)

pred_data$Mass_t1 <- pred_data$a * (pred_data$Size_t1 ^ pred_data$b) #mass after growth (t1)

# calculate difference t0 and t1
pred_data$Som_gain_b <- pred_data$Mass_t1 - pred_data$Mass

# gain is identical
all(pred_data$Som_gain_a == pred_data$Som_gain_b)
#> [1] TRUE

# calculate somatic loss
pred_data$Som_loss <- rfishprod::somaLoss(M = pred_data$Mort, Lmeas = pred_data$Size,
                                          a = pred_data$a, b = pred_data$b)

# calculate net changes
pred_data$Net_change_a <- pred_data$Som_gain_a - pred_data$Som_loss 

pred_data$Net_change_b <- pred_data$Som_gain_b - pred_data$Som_loss 

# check if any net change smaller than 0
which(pred_data$Net_change_a < 0)
#>  [1]  3  4  6  7  8 10 11 12 13 14 16 17 19 20 22 25 26 28 30

which(pred_data$Net_change_b < 0)
#>  [1]  3  4  6  7  8 10 11 12 13 14 16 17 19 20 22 25 26 28 30

# total net change
sum(pred_data$Net_change_a)
#> [1] -5.358749

sum(pred_data$Net_change_b)
#> [1] -5.358749
```

<sup>Created on 2021-06-18 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by renatoamorais** [comment]
Hi Max,

Thanks for raising this. You're not missing anything, at least not code-wise. This is purely a conceptual matter. Remember that productivity here is the balance between two functions.

One function, growth, more often than not follows an asymptotic trajectory (within the VBGM at least), decelerating near maximum sizes (or Linf). The other function, mortality, follows a negative exponential trajectory decelerating close to zero. The only thing you need to get negative values is that the estimates from the mortality function exceed the estimates from the growth function. 

So, when you calculate somatic mass loss from mortality (the probability of mortality times the individual mass), in effect you're assuming that individuals 'die' a little bit every single day. That assumption is, obviously, biologically nonsense. It surely does not make sense at the level of the individual. It may, however, be acceptable as a practicality at some aggregate levels (populations or communities, for instance), in specific situations. rfishprod:::repdata was generated randomly, independently for each vector and so it seems that these random fishes are 'dying' more than they are growing!

I would not recommend using somaLoss at all, unless you're prepared to carry all the assumptions it brings (you can read a bit more on the topic in the section "A framework for estimating fish productivity for high-diversity ecosystems" of the Coral Reefs paper, although that may not be in much more depth than already said here). 

Instead, I recommend using the probabilistic workflow: use predM to estimate the rate of decay of mortality probability and use applyMstoch to draw fate (survive or die) to each individual, productivity is the somatic growth of the surviving individuals only. Repeat this process hundreds of times.

Hope this helps, happy to keep up the discussion!

**Comment by mhesselbarth** [comment]
Hey, 

Thanks a lot for this comprehensive explanation.

That makes sense and is what I've expected. I've also tried the `somaLoss` code with some own data and had the same "problem" that fish "died" more than they grew. But as with the example data, I looked on an individual level, which given your explanation doesn't make too much sense. 

I think the probabilistic workflow makes much more sense on an individual level and I also like that it includes stochasticity.
